agents/postgres/pgstatsstatements/pgstatsstatements_test.go

```diff
@@ -0,0 +1,328 @@
+//			},
+//		}
+//		actual := makeBuckets(current, prev, logrus.WithField("test", t.Name()))
+//		require.Len(t, actual, 1)
+//		expected := &qanpb.MetricsBucket{
+//			Queryid:          "TruncateAndNew",
+//			Fingerprint:      "SELECT 'TruncateAndNew'",
+//			AgentType:        inventorypb.AgentType_QAN_MYSQL_PERFSCHEMA_AGENT,
+//			NumQueries:       5,
+//			MRowsAffectedCnt: 5,
+//			MRowsAffectedSum: 25,
+//		}
+//		assertBucketsEqual(t, expected, actual[0])
+//	})
+//}
+
+func logTable(t *testing.T, structs []reform.Struct) {
+	t.Helper()
+
+	if len(structs) == 0 {
+		t.Log("logTable: empty")
+		return
+	}
+
+	columns := structs[0].View().Columns()
+	var buf bytes.Buffer
+	w := tabwriter.NewWriter(&buf, 0, 0, 1, ' ', tabwriter.Debug)
+	_, err := fmt.Fprintln(w, strings.Join(columns, "\t"))
+	require.NoError(t, err)
+	for i, c := range columns {
+		columns[i] = strings.Repeat("-", len(c))
+	}
+	_, err = fmt.Fprintln(w, strings.Join(columns, "\t"))
+	require.NoError(t, err)
+
+	for _, str := range structs {
+		res := make([]string, len(str.Values()))
+		for i, v := range str.Values() {
+			res[i] = spew.Sprint(v)
+		}
+		fmt.Fprintf(w, "%s\n", strings.Join(res, "\t"))
+	}
+
+	require.NoError(t, w.Flush())
+	t.Logf("%s:\n%s", structs[0].View().Name(), buf.Bytes())
+}
+
+func setup(t *testing.T, db *reform.DB) *PGStatStatementsQAN {
+	t.Helper()
+
+	_, err := db.Exec(`do $$
+	begin
+	perform pg_stat_statements_reset();
+	end
+	$$;`)
+	require.NoError(t, err)
+
+	return newPgStatsStatementsQAN(db, "agent_id", logrus.WithField("test", t.Name()))
+}
+
+// filter removes buckets for queries that are not expected by tests.
+func filter(mb []*qanpb.MetricsBucket) []*qanpb.MetricsBucket {
+	res := make([]*qanpb.MetricsBucket, 0, len(mb))
+	for _, b := range mb {
+		switch {
+		// actions tests, MySQLVersion helper
+		case strings.HasPrefix(b.Fingerprint, "SHOW "):
+			continue
+		case strings.HasPrefix(b.Fingerprint, "ANALYZE "):
+			continue
+		case strings.HasPrefix(b.Fingerprint, "EXPLAIN "):
+			continue
+
+		// slowlog tests
+		case strings.HasPrefix(b.Fingerprint, "SELECT @@`slow_query_"):
+			continue
+
+		case strings.HasPrefix(b.Fingerprint, "SELECT @@`skip_networking`"):
+			continue
+
+		case strings.HasPrefix(b.Fingerprint, "TRUNCATE `performance_schema`"):
+			continue
+		case strings.HasPrefix(b.Fingerprint, "SELECT `performance_schema`"):
+			continue
+
+		default:
+			res = append(res, b)
+		}
+	}
+	return res
+}
+
+func TestPGStatStatementsQAN(t *testing.T) {
+	sqlDB := tests.OpenTestPostgreSQL(t)
+	defer sqlDB.Close() //nolint:errcheck
+	db := reform.NewDB(sqlDB, postgresql.Dialect, reform.NewPrintfLogger(t.Logf))
+
+	structs, err := db.SelectAllFrom(pgStatDatabaseView, "")
+	require.NoError(t, err)
+	logTable(t, structs)
+	structs, err = db.SelectAllFrom(pgStatStatementsView, "")
+	require.NoError(t, err)
+	logTable(t, structs)
+
+	engineVersionMajor, engineVersionMinor, engine := tests.PostgreSQLVersion(t, sqlDB)
```

Comment:
The second number is actually a patch version in later releases, not a minor one. Do we even have to return the second number for PostgreSQL? I don't think they do breaking changes like MySQL does in minor releases.

Reply:
so, they did it between 9.4 and 9.5

Reply:
😿 Let's return "9.5", "9.6", "10", "11", etc as a single string then.

Reply:
In that case I should write, all minor versions.
"10.0", "10.1", "10.2", "10.3", "10.4", "10.5", "10.6", "11.0" and e.t.c

Reply:
In 11.0 `0` is not a SemVer's minor version – it is SemVer's patch version. https://www.postgresql.org/support/versioning/ I don't think they do breaking change in patch versions (in "minor" versions since version 10). See also https://github.com/percona/pmm-agent/blob/70a48322e81eb991e1f21884d67037451409edb8/.travis.yml#L32-L34

Reply:
Minor releases are numbered by increasing the last part of the version number. Beginning with PostgreSQL 10, this is the second part of the version number, e.g. 10.0 to 10.1; for older versions this is the third part of the version number, e.g. 9.5.3 to 9.5.4.